src/common/libjob/job.c

```diff
@@ -658,6 +659,200 @@ int flux_job_id_encode (flux_jobid_t id,
+        return FLUX_JOB_RESULT_COMPLETED;
+    if (exception) {
+        if (exception_type != NULL) {
+            if (strcmp (exception_type, "cancel") == 0)
+                return FLUX_JOB_RESULT_CANCELED;
+            if (strcmp (exception_type, "timeout") == 0)
+                return FLUX_JOB_RESULT_TIMEOUT;
+        }
+    }
+    return FLUX_JOB_RESULT_FAILED;
+}
+
+static void result_eventlog_error_cb (flux_future_t *f, void *arg)
+{
+    json_t *res = arg;
+    json_t *result;
+    char *s = NULL;
+
+    if (flux_future_get (f, NULL) < 0 && errno != ENODATA) {
+        flux_future_continue_error (f, errno, NULL);
+        flux_future_destroy (f);
+        return;
+    }
+    if ((result = json_object_get (res, "result")))
+        json_integer_set (result, job_result_calc (res));
+    s = json_dumps (res, JSON_COMPACT);
+    flux_future_fulfill_next (f, s, free);
+    flux_future_destroy (f);
+}
+
+static int job_result_handle_exception (json_t *res,
+                                        json_t *context)
+{
+    json_t *type;
+    json_t *severity;
+    json_t *note = NULL;
+    json_t *exception = json_object_get (res, "exception");
+
+    if (!exception)
+        return -1;
+
+    if (json_unpack (context,
+                     "{s:O s:O s?:O}",
+                     "type", &type,
+                     "severity", &severity,
+                     "note", &note) < 0)
+        return -1;
+    if (json_integer_set (exception, 1) < 0
+        || json_object_set_new (res, "exception_type", type) < 0
+        || json_object_set_new (res, "exception_note", note) < 0
+        || json_object_set_new (res, "exception_severity", severity) < 0)
+        return -1;
+    return 0;
+}
+
+static void result_eventlog_cb (flux_future_t *f, void *arg)
+{
+    json_t *res = arg;
+    const char *entry = NULL;
+    const char *name = NULL;
+    json_t *context = NULL;
+    json_t *o = NULL;
+
+    if (flux_job_event_watch_get (f, &entry) < 0) {
+        /* This should never happen, since this is an "and_then" callback
+         */
+        goto error;
+    }
+    if (!(o = eventlog_entry_decode (entry))
+        || eventlog_entry_parse (o, NULL, &name, &context) < 0)
+        goto error;
+
+    if (!strcmp (name, "start")) {
+        json_t *started = json_object_get (res, "started");
+        if (started == NULL
+            || json_integer_set (started, 1) < 0)
+            goto error;
+    }
+    else if (!strcmp (name, "finish")) {
+        int status;
+        json_t *statusobj = json_object_get (res, "status");
+        json_t *exitcode = json_object_get (res, "exitcode");
+
+        if (statusobj == NULL
+            || exitcode == NULL
+            || json_unpack (context, "{s:i}", "status", &status) < 0
+            || json_integer_set (statusobj, status) < 0
+            || flux_job_event_watch_cancel (f) < 0)
+            goto error;
+
+        if (WIFSIGNALED (status))
+            json_integer_set (exitcode, WTERMSIG (status) + 128);
```

Comment:
Is `-WTERMSIG (status)` not a good option? That's what `FluxExecutor` does and also what you did [here](https://github.com/grondo/flux-core/commit/4a361896b1af0cc8a1d7b2bbc158135dde4dc151). It's fine either way of course, it would just be good to be consistent.

Reply:
Oh, sorry. This was adapted from existing code in `flux job status` which reports exit code in a way similar to the POSIX shell (128 + signal) if the function was terminated by a signal. I had forgotten about the `JobInfo` `returncode` property -- I think you are right, that is more useful here. We could also just add `returncode` in the Python interface.

Reply:
Is it feasible to just remove the `exitcode` JSON attribute and make the two interfaces construct the value that makes the most sense locally? So that `flux job status` could return 128 + signum and Python could return -signum.

Reply:
Yes, that is what I was thinking as well.

In fact, we could remove `exitcode`, rename `status` to `waitstatus` and `exception` to `exception_occurred` (and make it a boolean I think), and then the returned dict could be used to create a `JobInfo` object I think..